Test/dafny0/NoMoreAssume2Less2.dfy

```diff
@@ -0,0 +1,199 @@
+// RUN: %dafny "%s" > "%t"
+// RUN: %diff "%s.expect" "%t"
+
+// This file tests various ways of writing the equivalent of "assume false;" or "assert false;" inside a loop.
```

Comment:
What I find confusing about this PR is that it says "Boogie has behavior X, which we don't like, so we'll do something that happens to avoid X from triggering" without discussing why X is harming instead of helping us. Is behavior X bad or not? Can't we turn it off instead of avoid it?

Reply:
In Boogie, structured commands like `while` are just syntactic sugar for writing basic blocks and `goto`s. If you look at the desugared Boogie, then it is feature that Boogie cleans up the given control-flow graph by removing whatever follows an `assume false;`. For example, this can remove various join points in the control flow, which means that you need fewer single-assignment variables.

I had considered adding a feature to Boogie whereby you could annotate an `assume` or `assert` statement to say, "don't pay attention to whether or not the condition happens to be `false`". If Boogie had such an attribute, Dafny could use it in its translation of user-provided Dafny conditions to be checked. But this would require _both_ a new feature in Boogie _and_ a change to Dafny. So, it seemed simpler to just make a change in Dafny.

Two more notes, since I'm going on about this.

To make this change in Dafny, I had to cause anything that _would_ have looked like `assume false` (and similar for `assert`, `invariant`, `requires`, and `ensures`) to look like something else. One way to do this is to write `assume F(false)`, where `F` is an identity function (that is, a function that returns its argument). It turns out that Dafny was almost doing this already. I say "almost", because Dafny does translate the literal `false` into `Lit(false)`, where `Lit` is an identity function. For some reason that I don't understand, the Dafny verifier specifically looked for `Lit(...)` in the condition it was about to put into an `assume`, and then removed this `Lit` wrapper. My change, therefore, in Dafny was simplify to _not_ remove the `Lit` wrapper. This both makes Dafny simpler and addresses the problem.

Boogie has another pre-pass, which Dafny uses: abstract interpretation (`/infer:j`). It turns out that it also looks for `assume false`. Because of this abstract interpretation, it is actually possible (even after this PR) to still get the undesirable behavior of losing the back edge. See method `Conjuncts4` for an example. You have to use a particular combination of features to trip over this behavior, so I'm satisfied with what the current PR does. The current PR does fix all cases where _exactly_ a statement `assume false;` had caused back edges to dropped before, which is the common thing one adds during verification debugging.

Reply:
> If you look at the desugared Boogie, then it is feature that Boogie cleans up the given control-flow graph by removing whatever follows an assume false;. For example, this can remove various join points in the control flow, which means that you need fewer single-assignment variables.

Don't you just want to turn off this feature, which is effectively what you're doing with this PR but in a not 100% reliable way?

Reply:
Dafny generates many `assume false`s by itself. It is only for those that come from Dafny user input that we don't want Boogie's optimization.